### PR TITLE
Add opt-in strict action padding masking in GR00T N1.7

### PR DESCRIPTION
**Pull request by Henry-23**
GR00T N1.7 masks padded action values in the final loss, but padded noise and action tokens can still pass through the action encoder and DiT. This allows padding to affect predictions for valid action dimensions.

This PR adds an opt-in `strict_action_padding_mask` mode. When enabled, it applies the action mask before encoding, excludes padded horizon tokens from DiT self-attention, and keeps padded values zero throughout inference and Euler integration. The processor, PyTorch, ONNX, and TensorRT paths are updated consistently.

The option defaults to `False` so released checkpoints keep their existing behavior. New training runs may opt in when the same strict masking behavior will be used at inference. The setting is persisted in both the model and processor configs, and mismatched serving configurations fail fast.

The change does not add model parameters or alter checkpoint state dicts. TensorRT DiT engines built for strict mode need to include the new `hidden_attention_mask` input.

Validation:
- Valid outputs match between compact and padded sequences.
- Padded inputs have zero gradient in strict mode.
- Changing padded noise does not affect valid action loss in strict mode.
- The inference mask survives the processor-to-collator path.
- Model and processor strict-mask settings are checked during training and serving.

Refs #742

### CODE REVIEW

**Comment by jiajingc-nv** [comment]
Hi @Henry-23, I verified this branch against a held-out reproduction of #742. The diagnosis is right and the DiT-side fix (masking padded keys in self-attention, then re-zeroing) does close the leak on the training path.

One gap keeps it from taking effect at inference time. `get_action_with_features` gates every masking guard on `action_mask`, but the serving path never supplies it:

- `Gr00tPolicy._get_action` builds model inputs via `self.processor(messages)` (`gr00t/policy/gr00t_policy.py:409`), i.e. `Gr00tN1d7Processor.__call__`.
- `__call__` only produces `action_mask` on the branch that has an action, and inference steps carry none, so the key never reaches the batch.
- `process_observation`, the method this PR updates to emit the 3-D mask, has no production callers. That is also why `test_inference_action_mask_covers_horizon_and_dimension` stays green: it asserts correct behavior on a path the policy never takes.

So after this PR, training is fixed but a deployed policy still runs with padded horizon steps and padded action dims attending inside the DiT.

I opened Henry-23/Isaac-GR00T#1 against your branch with a fix and tests for it. It is rebased on aa03419 and merges cleanly, so you can pull it into this PR and keep authorship. Pls let me know your comments/suggestions if any, thanks!

**Comment by jiajingc-nv** [comment]
@ryhalabi flagging this one for your input, since it concerns the released checkpoint rather than just this PR.

### Where the released checkpoint's serving behavior changes

The model reasons over a fixed-size action canvas (`action_horizon` x `max_action_dim`). A real embodiment fills only a corner of it; the rest is padding. Every checkpoint shipped so far was trained with that padding present, and has been serving with it present too. The model has never operated without it.

#741 takes the padding away. For a model retrained under #741 that is strictly correct: training and serving agree, and neither sees padding. For an already-released checkpoint it is a distribution shift in the opposite direction, since we would be hiding something the model has always seen.

That was moot until now, because #741's inference-side masking never actually ran: the live path never supplies `action_mask` (details in my comment above). Closing that gap is exactly what Henry-23/Isaac-GR00T#1 does. So that follow-up, not #741 on its own, is the point at which the released checkpoint's serving behavior changes.

The shift is not marginal. `nvidia/GR00T-N1.7-3B` and `nvidia/GR00T-N1.7-LIBERO` both ship `action_horizon: 40` and `max_action_dim: 132`, while `libero_sim` declares `delta_indices=list(range(16))` over 7 action keys. For that embodiment:

- 24 of 40 horizon slots are padding. Because `action_token_mask = action_mask.any(dim=-1)`, those 24 tokens drop out of DiT self-attention entirely. That is 60% of the action tokens, for a model trained with all 40 attending.
- 125 of 132 action dims are padding (94.7%), zeroed before the action encoder.

Suggested handling, either one:

- evaluate the released N1.7 checkpoints before merging and report the delta; or
- land the training-side fix now and keep the serving-side masking opt-in until a checkpoint retrained under #741 exists.

I cannot assess the quality impact myself. Could you pls help with this? Thanks a lot!

**Comment by Henry-23** [comment]
Hi @jiajingc-nv, thanks for raising the compatibility concern. 

I’ve updated the PR so strict action-padding masking is opt-in, while legacy behavior remains the default for released checkpoints.

The setting is persisted in both the model and processor configs and kept consistent across training and inference. I also added fail-fast checks to prevent strict mode from silently running without an action mask. This avoids changing the default behavior of released checkpoints while providing an opt-in path for consistent masking during training and inference.

**Comment by jiajingc-nv** [comment]
@Henry-23 @ryhalabi Inspired by this thread, I found and fixed another bug [here](https://github.com/NVIDIA/Isaac-GR00T/pull/744). Pls take a look if you're interested.

**Comment by jiajingc-nv** [comment]
@Henry-23 Re-reviewed the current head. The opt-in redesign resolves the checkpoint-compatibility concern cleanly — defaults are `False` in `Gr00tN1d7Config`, `FinetuneConfig` and the processor, `from_pretrained` sets the same default, and `validate_strict_action_padding_mask` fails fast when the two sides disagree — so a released checkpoint's serving behavior is bit-identical until someone opts in. `test_collated_inference_batch_carries_action_mask` closes the processor→collator seam, and `test_masked_padding_matches_compact_sequence` covers the padded case that the all-valid no-op test could not.

One retraction: my earlier note that `if action_mask is not None:` in `__call__` had become a dead branch is obsolete. With the flag off the mask is legitimately `None`, so that branch is live again — please leave it.

Two things I'd still suggest a follow-up commit for, plus two nits.

### 1. The two mask derivations differ, and the obvious cleanup is the harmful one

Training derives the mask extent from the data, inference from the config:

```
processing_gr00t_n1d7.py:636   action_dim     = normalized_actions.shape[1]
processing_gr00t_n1d7.py:649   action_horizon = normalized_actions.shape[0]
_build_action_mask       :468  action_dim     = self.state_action_processor.get_action_dim(...)
_build_action_mask       :467  action_horizon = len(action_config.delta_indices)
```

They cannot disagree today: `gr00t/data/stats.py` clamps the sampling range with `usable_length = len(df) - delta_indices[-1]`, so a chunk is always full length. That makes this a latent invariant rather than a live bug — but nothing in the tree records it. Sooner or later someone will spot the two spellings and unify them, and unifying onto the config-derived side is actively harmful: on a short chunk it would mark zero-padded targets as valid and ask the loss to fit those zeros. Keeping the training branch data-derived is the right call; a sentence in `_build_action_mask`'s docstring saying so, and why, would stop the "cleanup".

Related: `test_action_mask_identical_with_and_without_action` uses a full-length chunk, so it pins the two paths only at the point where the clamp already guarantees they agree.

### 2. `_expand_action_mask`'s `ndim == 2` branch no longer has a producer

Every producer now reaches the head as 3-D: `_build_action_mask` and the training `ones_like` are `(H, D)` per sample and collate to `(B, H, D)`, and `process_observation` repeats to `(B, H, D)` directly. So the only tensor that can enter the `ndim == 2` branch is a `(B, H)` mask, which can only come from a bug — and it gets `unsqueeze`d and broadcast into a valid-looking input rather than raising. Dropping the branch and letting the shape check reject 2-D would turn that into a loud failure.

### Nits

- `_build_action_mask`: `if action_horizon > 0 and action_dim > 0:` guards a zero-length slice assignment, which is already a no-op.
- `test_action_mask_identical_with_and_without_action`: `torch.equal` fits a 0/1 mask better than `assert_close`. (`assert_close` in the DiT no-op test is right — the SDPA-backend rationale there is well put.)

**Comment by Henry-23** [comment]
@jiajingc-nv I’ve pushed a follow-up that documents why training derives the mask extent from the observed action tensor while inference uses the embodiment config. I also removed the obsolete 2-D mask broadcast so malformed horizon-only masks now fail loudly, added a regression test, and addressed the two smaller nits. The action_mask is not None branch remains unchanged.
